public/app/features/dashboard/state/reducers.ts

```diff
@@ -1,21 +1,46 @@
-import { DashboardState } from 'app/types';
-import { Action, ActionTypes } from './actions';
+import { DashboardState, DashboardLoadingState } from 'app/types/dashboard';
+import { loadDashboardPermissions, setDashboardLoadingState, setDashboardModel, setDashboardLoadingSlow } from './actions';
+import { reducerFactory } from 'app/core/redux';
 import { processAclItems } from 'app/core/utils/acl';
 
 export const initialState: DashboardState = {
+  loadingState: DashboardLoadingState.NotStarted,
+  isLoadingSlow: false,
+  model: null,
   permissions: [],
 };
 
-export const dashboardReducer = (state = initialState, action: Action): DashboardState => {
-  switch (action.type) {
-    case ActionTypes.LoadDashboardPermissions:
-      return {
-        ...state,
-        permissions: processAclItems(action.payload),
-      };
-  }
-  return state;
-};
+export const dashboardReducer = reducerFactory(initialState)
+  .addMapper({
+    filter: loadDashboardPermissions,
+    mapper: (state, action) => ({
+      ...state,
+      permissions: processAclItems(action.payload),
+    }),
+  })
+  .addMapper({
+    filter: setDashboardLoadingState,
+    mapper: (state, action) => ({
+      ...state,
+      loadingState: action.payload
+    }),
+  })
+  .addMapper({
+    filter: setDashboardModel,
+    mapper: (state, action) => ({
+      ...state,
+      model: action.payload,
+      isLoadingSlow: false,
```

Comment:
Are there any other actions (i.e. when something goes wrong) that should reset isLoadingSlow?

Reply:
Good feedback, this caused me to completely rethink the actions :)